[receiver/kafkareceiver]: allow tunable fetch sizes

cxdy opened the pull request:
**Description:**
This commit adds the ability to tune the minimum, default and maximum fetch sizes for the Kafka Receiver in the OpenTelemetry configuration file.

The defaults are kept consistent with the defaults imposed by [sarama](https://pkg.go.dev/github.com/shopify/sarama#Config)

**Link to tracking Issue:** 
Resolves https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/22741

**Testing:**
Built the image with `make docker-otelcontribcol`, uploaded to [Docker Hub](https://hub.docker.com/r/ckaczynski/otelcontribcol) to pull the image into our Kubernetes deployment, and hit the cluster with 200k logs per second with the following settings:
```
        default_fetch_size: 15728640 # 15MB
        max_fetch_size: 31457280 # 30MB
        min_fetch_size: 1048576  # 1MB
```
These limits are absurdly high for no reason other than because I just wanted to set them higher and see how far I could push it in our cluster.

No new tests were added, but added to existing configuration tests in config.go and factory.go

**Documentation:** 
Added the new configuration options to the receiver's README

Commit:
[receiver/kafkaexporter]: allow tunable fetch sizes

This commit adds the ability to tune the minimum, default and maximum fetch sizes for Kafka in the OpenTelemetry configuration file.

cxdy committed:
Merge pull request #1 from cxdy/feat/kafkareceiver-fetch-configurable

[receiver/kafkareceiver]: allow tunable fetch sizes

cxdy committed:
Merge branch 'main' into main

cxdy committed:
Merge branch 'main' into main

cxdy committed:
Merge branch 'main' into main

cxdy commented:
Sorry about the merge commits, can probably squash them at merge but I can rebase too.

cxdy committed:
Merge branch 'main' into main

cxdy commented:
Ah, apologies, missed the changelog entry. Will add shortly.

Commit:
chore: add changelog

cxdy committed:
Merge pull request #3 from cxdy/feat/kafkareceiver-fetch-configurable

chore: add changelog

Commit:
fix: tests using wrong int type

cxdy committed:
Merge pull request #4 from cxdy/feat/kafkareceiver-fetch-configurable

fix: tests using wrong int type

cxdy commented:
Unit tests _should_ pass now, I didn't realize I needed to specify `int32` for the type in `factory_test.go`.

Commit:
fix: specify int32

cxdy committed:
Merge pull request #5 from cxdy/feat/kafkareceiver-fetch-configurable

fix: specify int32

mwear reviewed (approved):
LGTM. Could the codeowners take a look cc: @pavolloffay @MovieStoreGuy.

iblancasa commented on receiver/kafkareceiver/config.go:
```diff
@@ -78,6 +78,13 @@ type Config struct {
+	MinFetchSize int32 `mapstructure:"min_fetch_size"`
+	// The default bytes per fetch from Kafka (default "1048576")
+	DefaultFetchSize int32 `mapstructure:"default_fetch_size"`
+	// The maximum bytes per fetch from Kafka (default "0")
```
```suggestion
	// The maximum bytes per fetch from Kafka (default "0", no limit)
```

cxdy committed:
fix: unnecessary conversion

 Error: kafkareceiver/factory_test.go:33:23: unnecessary conversion (unconvert)
501
	assert.Equal(t, int32(defaultMaxFetchSize), int32(cfg.MaxFetchSize))

cxdy committed:
docs: clarify max limit (none)

Co-authored-by: Israel Blancas <iblancasa@gmail.com>

cxdy committed:
Merge branch 'main' into main

cxdy committed:
Merge branch 'main' into main

cxdy committed:
Merge branch 'main' into main

cxdy committed:
Merge branch 'main' into main

MovieStoreGuy commented:
Can I ask you to fix up the conflicts?

cxdy committed:
Merge branch 'main' into main

cxdy commented:
Sure can!

MovieStoreGuy commented on receiver/kafkareceiver/factory_test.go:
```diff
@@ -28,6 +28,9 @@ func TestCreateDefaultConfig(t *testing.T) {
 	assert.Equal(t, defaultInitialOffset, cfg.InitialOffset)
 	assert.Equal(t, defaultSessionTimeout, cfg.SessionTimeout)
 	assert.Equal(t, defaultHeartbeatInterval, cfg.HeartbeatInterval)
+	assert.Equal(t, int32(defaultMinFetchSize), cfg.MinFetchSize)
```
``suggestion
	assert.Equal(t, defaultMinFetchSize, cfg.MinFetchSize)
	assert.Equal(t, defaultDefaultFetchSize, cfg.DefaultFetchSize)
	assert.Equal(t, defaultMaxFetchSize, cfg.MaxFetchSize)
```

To fix up the unneeded conversions.

MovieStoreGuy commented on receiver/kafkareceiver/factory_test.go:
```diff
@@ -28,6 +28,9 @@ func TestCreateDefaultConfig(t *testing.T) {
 	assert.Equal(t, defaultInitialOffset, cfg.InitialOffset)
 	assert.Equal(t, defaultSessionTimeout, cfg.SessionTimeout)
 	assert.Equal(t, defaultHeartbeatInterval, cfg.HeartbeatInterval)
+	assert.Equal(t, int32(defaultMinFetchSize), cfg.MinFetchSize)
+	assert.Equal(t, int32(defaultDefaultFetchSize), cfg.DefaultFetchSize)
+	assert.Equal(t, int32(defaultMaxFetchSize), cfg.MaxFetchSize)
```
```suggestion
	assert.Equal(t, defaultMinFetchSize, cfg.MinFetchSize)
	assert.Equal(t, defaultDefaultFetchSize, cfg.DefaultFetchSize)
	assert.Equal(t, defaultMaxFetchSize, cfg.MaxFetchSize)
```

To fix up the unneeded conversions.

MovieStoreGuy reviewed (approved):
Just need to fix up conversions.

cxdy committed:
fix: remove unnecessary conversions

Co-authored-by: Sean Marciniak <30928402+MovieStoreGuy@users.noreply.github.com>

cxdy commented:
Thanks for that, haven't had a chance to grab my laptop yet. Committed!

cxdy committed:
Merge branch 'main' into main
